### ISSUE TITLE
Text rendering has glitches for pathologically large files

### ISSUE THREAD

**Issue by imron** [zed-industries/feedback#107]
**Describe the bug**
When opening a large file (e.g. greater several hundred MB), the text rendering of lines starts to overlap towards the end of the file.  In the example given below, this starts to happen ~300,000 lines in, where the cursor is rendered slightly offset the line highlight, and the problem gets progressively worse by the end of the file, where lines completely overlap each other (but not all lines).

For the reproduction steps below, this image shows the last line that has a cursor aligned with the height of the line highlight

<img width="799" alt="good-cursor" src="https://user-images.githubusercontent.com/1484640/174710826-c0a19957-4358-4c3f-89c3-e7f4d99f7559.png">

These next two images show the first two lines that have a mis-aligned cursor:

<img width="790" alt="bad-cursor-1" src="https://user-images.githubusercontent.com/1484640/174710933-279f5160-7016-40f8-9282-b4ce5a606534.png">

<img width="847" alt="bad-cursor-2" src="https://user-images.githubusercontent.com/1484640/174709308-05d5bd10-f503-4fe1-8536-fbadd43613fb.png">

And eventually things get so bad that text just overlaps itself.

<img width="822" alt="overlapping-lines" src="https://user-images.githubusercontent.com/1484640/174709690-14bb6d87-b9b2-426d-b380-b6aed538da54.png">

Note also that the highlighted line number is at a different location from the cursor and the line selection highlight.

**To Reproduce**
1. Generate a large file.  For this example, I downloaded ["Moby Dick" from the Gutenberg Project](https://www.gutenberg.org/files/2701/2701-0.txt), saved it as moby-dick.txt and `cat`ed it multiple times to get a 1.2GB file

       for i in $(seq 1 1000); do cat moby-dick.txt >> moby-dick-1GB.txt; done 

   Note: a text file was used to avoid issues with syntax highlighting and large files.

2. Open the file in Zed (and wait a minute for it to load). 
3. Control-G and go to line 2796200 (note the cursor is incorrectly offset at all lines after this line, but correctly offset on all lines before it).
4. Command-DownArrow to go the end of the file.
5. Note that some text lines overlap each other.

**Expected behavior**
Text should render correctly, and cursors, line-numbers and line highlights should all align with each other regardless of the size of the underlying file. 

Opening large files like this is definitely not going to be common, however any editor that claims to be fast should handle pathological cases like this correctly because it's the first thing people are going to do to see how well those claims hold up 😄

**Screenshots**
See above.

**Environment:**
 - Architecture [Intel]:
 - Mac OS Version: [macOS 12.3.1]:
 - Zed Version [e.g. 0.39.0]:

**Comment by JosephTLyons** [comment]
Wonder if it is related at all to https://github.com/zed-industries/feedback/issues/60

**Comment by imron** [comment]
I think it likely is (I realised after posting that I should've searched existing issues first).

Anyway, if you increase the size of the file you used in #60 by 10-fold, and add a few longer lines, I'm sure you'll see the same thing.

**Comment by imron** [comment]
I wonder if it's something like the line-height being a non-integral value e.g. 48.000000357 which is being clamped to 48 for some things but not others, and then at some point the line height is multiplied by line number.  For large enough values of line number 48 * line number vs 48.000000357 * line number is enough to start causing pixel offsets.

**Comment by iamnbutler** [comment]
Thanks for exploring this! Yeah, as Joseph mentioned we are somewhat aware of this. 

We'll need to figure out where those non-clamped values are, but it's possible we'll update the system we use to render text in the buffer first I imagine (to allow us to render elements that are arbitrary in height – currently all elements in the buffer need to be multiples of the buffer line height.)

I love that people are trying to break this :D

Have you found a file large enough to make the editor lag significantly/hang?

**Comment by imron** [comment]
> Have you found a file large enough to make the editor lag significantly/hang?

The 1GB file I talk about above takes about a minute to load and has spinning beach balls from about 5 seconds in.  It also takes up 9GB of RAM when open (according to Activity Monitor) from a starting value of 135MB.

Closing the file (with Zed still open) drops memory usage to 309 MB.  Reopening and closing (and reopening and closing) brings this up to 900MB - initially I thought this might be a leak, but it caps out at 900MB even after multiple opens and closes so maybe it's some sort of pool rather than a leak.

The 2GB version is similar but takes longer to load and also uses 18GB of RAM ,with 773MB in use after closing the file first time (didn't try multiple reloads yet). 

The 100GB version stops loading after a minute and control returns to Zed but no file is displayed - I guess there is an error and Zed bails on loading the file, but at least Zed stays up.  Memory usage however is high at 32GB even though nothing is open.

Once loaded though, navigation of large files is mostly without lag (trackpad scrolling upwards from the end of the file sometimes causes a slight pause), but jumping to line numbers works instantly.

> I love that people are trying to break this :D

I write a tool that has a file viewer, and have a bunch of files ready to go of pathological cases that cause issues for most other viewers/editors (but not mine 😄).

Zed has handled the pathological cases pretty well so far - with the exception of slow loading, high memory usage, and this rendering glitch when dealing with _extremely_ large files.

The pathological cases are mostly focused on regular text though - things like really long single lines of text, unicode outside the BMP, MBs of nothing but thin characters e.g. `i` (which not so much an issue in monospaced fonts and I haven't investigated how to change fonts in Zed yet) and then just really large files.

I'll spend a bit of time seeing if I can come up with some pathological files containing source code that mess with things like syntax highlighting and the like.

**Comment by JosephTLyons** [comment]
I recall an earlier version of Zed entirely froze while trying to use Go To Line within a massive file - I had to kill the instance and reopen it.

**Comment by JosephTLyons** [comment]
I started thinking about this again today and it sparked me to question: "At what point is it good enough?"  For every large file that the devs optimize for, a larger one can be made.  I'm all for having the fastest editor around, but it seems like some sort of line has to be drawn somewhere; defining that line seems extremely difficult.

**Comment by imron** [comment]
> For every large file that the devs optimize for, a larger one can be made

This is true, however things reach a point where designing things to work well with 1GB files usually means you can handle files of size N GB without much issue (at least this was true for the file viewer I wrote).

Even if that's not the case, and it can't be made fast because of data-structures required for other more important functionality, the editor still shouldn't show rendering glitches.  That is, it should be slow but correct, rather than fast but incorrect.

**Comment by JosephTLyons** [comment]
> > For every large file that the devs optimize for, a larger one can be made
> 
> This is true, however things reach a point where designing things to work well with 1GB files usually means you can handle files of size N GB without much issue (at least this was true for the file viewer I wrote).
> 
> Even if that's not the case, and it can't be made fast because of data-structures required for other more important functionality, the editor still shouldn't show rendering glitches. That is, it should be slow but correct, rather than fast but incorrect.

That is a fair point - that things should render correctly, even if something is talking a long time to process.

**Comment by b-fuze** [comment]
I tried it on the whole Linux kernel source code (1.2GB) as a single file just to see how it would handle syntax highlighting + the large file. It seems that Zed gave up on syntax highlighting the file? Not sure.

I checked out the Linux kernel at revision `8ed710da2873c2aeb3bb805864a699affaf1d03b` and ran the following command to produce the C file:
```bash
find -type f -name '*.h' -o -name '*.c' | while read file; do cat "$file"; done > large-c-file.c
```

https://user-images.githubusercontent.com/10504957/206196971-a2bceca7-5c0b-45da-938f-4d64d34bb6c0.mp4

**Comment by JosephTLyons** [comment]
@b-fuze, just curious, do you know which editors (GUI or terminal-based) might do well in handling that file?  I'm curious to know which editors are handling these massive files better than Zed currently is.

**Comment by b-fuze** [comment]
@JosephTLyons interestingly, VSCode opens the file a lot faster than Zed does and can display the lines at the bottom just fine. I understand that this is an edge case and I am not saying that this should be a supported use-case, nor do I expect it to be. I would say that reaching VSCode's time for opening would be ideal though. i.e. Don't wait a whole minute freezing the editor with the spinning beach ball to open it. The line rendering issue should probably be solved as well, but it's not as important imo as not freezing the editor for a long time when opening it. I'd actually be okay with it giving a message saying that it can open it but it'll take a while as a stop gap measure. 

The only reason I tested this was because I'm really curious how the syntax highlighting performs, and other large files (much smaller than this one though) Zed actually does a phenomenal job with the syntax highlighting. So I was curious how it would handle a very large C file that isn't just some repeated text :smile: 

I don't actually see any real problem here other than notifying the user of what will happen before doing it. Here's a screenshot of VSCode though:
![image](https://user-images.githubusercontent.com/10504957/206203172-d8558ad8-1e40-4c28-8903-a01a224d35b6.png)

**Comment by imron** [comment]
> I tried it on the whole Linux kernel source code (1.2GB) as a single file just to see how it would handle syntax highlighting + the large file.

I love that you tried this.

>  I'm curious to know which editors are handling these massive files better than Zed currently is.

For me, part of it is not so much which editors are better, but rather pushing the limits to see what Zed is capable of.

These cases are unlikely to be encountered in most real world usage (although I know of people who have sometimes needed to edit multi-gb json and/or csv files).
